From: =?UTF-8?q?Cristian=20Rodr=C3=ADguez?= <crrodriguez@opensuse.org>
Date: Fri, 28 Feb 2014 17:28:13 +0100
Subject: [PATCH] Switch to sys/xattr.h

Since a very long time (over ten years) the xattrs functions
that libacl uses have been provided by libc.
This commit switches the code to use libc and leave libxattr
behind.
---
 configure.ac            |  4 ++--
 libacl/Makemodule.am    |  4 ++--
 libacl/error_context.h  | 53 +++++++++++++++++++++++++++++++++++++++++++++++++
 libacl/perm_copy_fd.c   |  7 +------
 libacl/perm_copy_file.c |  7 +------
 5 files changed, 59 insertions(+), 16 deletions(-)
 create mode 100644 libacl/error_context.h

diff --git a/configure.ac b/configure.ac
index 4abb9933948a..5915ee3e492c 100644
--- a/configure.ac
+++ b/configure.ac
@@ -51,8 +51,8 @@ fi
 LT_REVISION=$(printf "%d%d%02d" "$1" "$2" "$3")
 AC_SUBST(LT_REVISION)
 
-AC_PACKAGE_NEED_ATTR_ERROR_H
-AC_PACKAGE_NEED_GETXATTR_LIBATTR
+AC_CHECK_HEADERS([sys/xattr.h],,
+                 [AC_MSG_ERROR([sys/xattr.h header not found])])
 
 AC_CONFIG_COMMANDS([include/acl],
 	[dnl
diff --git a/libacl/Makemodule.am b/libacl/Makemodule.am
index c35214c7d018..64b70b40951a 100644
--- a/libacl/Makemodule.am
+++ b/libacl/Makemodule.am
@@ -74,8 +74,8 @@ INTERNAL_CFILES = \
 
 libacl_la_DEPENDENCIES = exports libmisc.la
 libacl_la_SOURCES = $(CFILES) $(HFILES)
-libacl_la_LIBADD = -lattr libmisc.la
+libacl_la_LIBADD = libmisc.la
 libacl_la_CFLAGS = $(AM_CFLAGS) -include $(srcdir)/libacl/perm_copy.h
-libacl_la_LDFLAGS = \
+libacl_la_LDFLAGS = -no-undefined \
 	-Wl,--version-script,$(top_srcdir)/exports \
 	-version-info $(LTVERSION)
diff --git a/libacl/error_context.h b/libacl/error_context.h
new file mode 100644
index 000000000000..fa383f4cb5db
--- /dev/null
+++ b/libacl/error_context.h
@@ -0,0 +1,53 @@
+/*
+  Copyright (C) 2009  Andreas Gruenbacher <agruen@suse.de>
+
+  This program is free software: you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation, either version 2.1 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef __ERROR_CONTEXT_T
+#define __ERROR_CONTEXT_T
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct error_context {
+       /* Process an error message */
+       void (*error) (struct error_context *, const char *, ...);
+
+       /* Quote a file name for including in an error message */
+       const char *(*quote) (struct error_context *, const char *);
+
+       /* Free a quoted name */
+       void (*quote_free) (struct error_context *, const char *);
+};
+
+#ifdef ERROR_CONTEXT_MACROS
+# define error(ctx, args...) do { \
+       if ((ctx) && (ctx)->error) \
+               (ctx)->error((ctx), args); \
+       } while(0)
+# define quote(ctx, name) \
+       ( ((ctx) && (ctx)->quote) ? (ctx)->quote((ctx), (name)) : (name) )
+# define quote_free(ctx, name) do { \
+       if ((ctx) && (ctx)->quote_free) \
+               (ctx)->quote_free((ctx), (name)); \
+       } while(0)
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif  /* __ERROR_CONTEXT_T */
diff --git a/libacl/perm_copy_fd.c b/libacl/perm_copy_fd.c
index f9ae690918fa..8e1eb939512b 100644
--- a/libacl/perm_copy_fd.c
+++ b/libacl/perm_copy_fd.c
@@ -17,9 +17,8 @@
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
 
-#if defined (HAVE_CONFIG_H)
 #include "config.h"
-#endif
+
 #if defined(HAVE_LIBACL_LIBACL_H)
 # include "libacl.h"
 #endif
@@ -39,11 +38,7 @@
 #endif
 
 #define ERROR_CONTEXT_MACROS
-#ifdef HAVE_ATTR_ERROR_CONTEXT_H
-#include <attr/error_context.h>
-#else
 #include "error_context.h"
-#endif
 
 #if !defined(ENOTSUP)
 # define ENOTSUP (-1)
diff --git a/libacl/perm_copy_file.c b/libacl/perm_copy_file.c
index 023294dc5335..9d314a5c8d36 100644
--- a/libacl/perm_copy_file.c
+++ b/libacl/perm_copy_file.c
@@ -17,9 +17,8 @@
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
 
-#if defined (HAVE_CONFIG_H)
 #include "config.h"
-#endif
+
 #if defined(HAVE_LIBACL_LIBACL_H)
 # include "libacl.h"
 #endif
@@ -39,11 +38,7 @@
 #endif
 
 #define ERROR_CONTEXT_MACROS
-#ifdef HAVE_ATTR_ERROR_CONTEXT_H
-#include <attr/error_context.h>
-#else
 #include "error_context.h"
-#endif
 
 #if !defined(ENOTSUP)
 # define ENOTSUP (-1)
